src/NLog/Targets/WebServiceTarget.cs

```diff
@@ -199,6 +200,162 @@ public WebServiceTarget(string name) : this()
+                : logEvents.BucketSort(convetToParameterArrayDelegate, ArrayDeepEqualityComparer<object>.Default);
+            foreach (var bucket in parameterBuckets)
+            {
+                for (int i = 0; i < this.Parameters.Count; ++i)
+                {
+                    var param = this.Parameters[i];
+                    if (param.GroupLayoutActive)
+                    {
+                        bucket.Key[i] = ConvertParameterGroupValue(bucket.Value, param);
+                    }
+                }
+
+                if (bucket.Value.Count > 1)
+                {
+                    AsyncContinuation[] groupContinuations = new AsyncContinuation[bucket.Value.Count];
+                    for (int i = 0; i < groupContinuations.Length; ++i)
+                    {
+                        groupContinuations[i] = bucket.Value[i].Continuation;
+                    }
+                    AsyncContinuation groupCompleted = (ex) =>
+                    {
+                        for (int i = 0; i < groupContinuations.Length; ++i)
+                            try { groupContinuations[i].Invoke(ex); } catch { /* Nothing to do about it */ };
+                    };
+                    DoGroupInvokeAsync(headerValues, bucket.Key, groupCompleted);
+                }
+                else
+                {
+                    DoGroupInvokeAsync(headerValues, bucket.Key, bucket.Value[0].Continuation);
+                }
+            }
+        }
+        private SortHelpers.KeySelector<AsyncLogEventInfo, object[]> convetToParameterArrayDelegate;
+
+        class ArrayDeepEqualityComparer<TValue> : IEqualityComparer<TValue[]>
+        {
+            public static readonly ArrayDeepEqualityComparer<TValue> Default = new ArrayDeepEqualityComparer<TValue>();
+
+            public bool Equals(TValue[] x, TValue[] y)
+            {
+                if (x.Length != y.Length)
+                    return false;
+
+                object xval, yval;
+                for (int i = 0; i < x.Length; ++i)
+                {
+                    xval = x[i];
+                    yval = y[i];
+                    if (xval != null && yval != null)
+                    {
+                        if (!x[i].Equals(y[i]))
+                            return false;
+                    }
+                    else if (!ReferenceEquals(xval, yval))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            public int GetHashCode(TValue[] obj)
+            {
+                int hashCode = obj.Length.GetHashCode();
+                for (int i = 0; i < obj.Length; ++i)
+                {
+                    if (obj[i] != null)
+                        hashCode = hashCode ^ obj[i].GetHashCode();
```

Comment:
AFAIK it good practice to use a prime number, see for example http://www.aaronstannard.com/overriding-equality-in-dotnet/

Ps. Why aren't you using R#? I'm convinced that your good enough to skip implementing simple stuff. :)